src/playlist-controller.js

```diff
@@ -2104,51 +2106,83 @@ export class PlaylistController extends videojs.EventTarget {
+      this.contentSteeringController_.addAvailablePathway(this.pathwayAttribute_(playlist));
+    }
+    this.contentSteeringController_.assignTagProperties(main.uri, main.contentSteering);
+    // request the steering manifest immediately if queryBeforeStart is set.
+    if (this.contentSteeringController_.queryBeforeStart) {
+      // When queryBeforeStart is true, initial request should omit steering parameters.
+      this.contentSteeringController_.requestSteeringManifest(true);
+      return;
+    }
+    // otherwise start content steering after playback starts
+    this.tech_.one('canplay', () => {
+      this.contentSteeringController_.requestSteeringManifest();
+    });
+  }
 
-    const updateSteeringValues = (main) => {
-      for (const playlist of main.playlists) {
-        this.contentSteeringController_.addAvailablePathway(this.pathwayAttribute_(playlist));
-      }
-
-      this.contentSteeringController_.assignTagProperties(main.uri, main.contentSteering);
-    };
-
-    updateSteeringValues(initialMain);
+  /**
+   * Reset the content steering controller and re-init.
+   */
+  resetContentSteeringController_() {
+    this.contentSteeringController_.clearAvailablePathways();
+    this.contentSteeringController_.dispose();
+    this.initContentSteeringController_();
+  }
 
+  /**
+   * Attaches the listeners for content steering.
+   */
+  attachContentSteeringListeners_() {
     this.contentSteeringController_.on('content-steering', this.excludeThenChangePathway_.bind(this));
-
-    // We need to ensure we update the content steering values when a new
-    // manifest is loaded in live DASH with content steering.
     if (this.sourceType_ === 'dash') {
-      this.mainPlaylistLoader_.on('mediaupdatetimeout', () => {
-        this.mainPlaylistLoader_.refreshMedia_(this.mainPlaylistLoader_.media().id);
-
-        // clear past values
-        this.contentSteeringController_.abort();
-        this.contentSteeringController_.clearTTLTimeout_();
-        this.contentSteeringController_.clearAvailablePathways();
-
-        updateSteeringValues(this.main());
-      });
-    }
+      this.mainPlaylistLoader_.on('loadedplaylist', () => {
+        const main = this.main();
+        // check if steering tag or pathways changed.
+        const didDashTagChange = this.contentSteeringController_.didDASHTagChange(main.uri, main.contentSteering);
+        const didPathwaysChange = () => {
+          const availablePathways = this.contentSteeringController_.getAvailablePathways();
+          const newPathways = [];
+
+          for (const playlist of main.playlists) {
+            const serviceLocation = playlist.attributes.serviceLocation;
+
+            if (serviceLocation) {
+              newPathways.push(serviceLocation);
+              if (!availablePathways.has(serviceLocation)) {
+                return true;
+              }
+            }
+            if (!newPathways.length && availablePathways.size) {
+              return true;
```

Comment:
I think I am having a tough time understanding what this if statement is doing.. This will always return false as long as the first playlist has a `serviceLocation` property, right? Do we need to check when `availablePathways` has a value that `newPathways` doesn't, AKA the removal of a pathway (I think).

Reply:
My mistake. This is supposed to be outside of the loop. This would be the case where a manifest refreshes with no `serviceLocation`s, represented by the `newPathways` collection. If the `.mpd` refreshes with no `serviceLocations` however the steering controller has available pathways, we want to return true and signal a change.

Reply:
I think that makes more sense. I might be overthinking this, but won't `availablePathways.size` almost always return true? I think in this case we would want to return false.

Example scenario: In a case where nothing changes, newPathways length = 0, and `availablePathways` stays the same, but has a length. I think in this case we should be returning false, but is returning true currently.

Let me know if I am off here, as I definitely could be.

Reply:
Since we add a `serviceLocation` to `newPathways` regardless of `availablePathways` having the pathway or not (only if it exists), the example scenario would actually indicate changes and you would want to return `true` I think. If `newPathways.length === 0` then the main manifest/.mpd update has no `serviceLocation`s, so if the content steering controller has `availablePathways` and no `newPathway`s exist in the main manifest after a refresh, we would be changing state and need to disable content steering as the current main manifest has no `serviceLocation`s to steer with.

Reply:
Thanks for the explanation, that makes more sense!